### PR TITLE
wip: Model inheritance

### PR DESCRIPTION
**Pull request by jgibson**
This adds support for tracking inheritance of fields and values to shr-expand. This adds another pass after the expansion to compare the fields and values of each child element against its parents. If a field or value exists on at least one parent and it is equal to the child's copy then it is considered `inherited`. If it is not equal to any parent's copy then it is considered `overridden`. If the field or value does not exist on the parent then it is considered to be new and the `inheritance` field is left null/undefined.

Note that AFAICT the none of the spec uses multiple inheritance (and I don't think that any of our tests really exercise it) so I'm not sure how it will behave when using that language feature.

One outstanding issue is the inheritance of Entry fields. I think it is because the expander special-cases inheritance from shr.base.Entry in findMatchInDataElement().

### CODE REVIEW

File: lib/expand.js [hunk 1]
```diff
@@ -43,6 +43,107 @@ class Expander {
...
+        'shr.core': { 'Version': true },
+        'shr.base': {}
+      };
+      for (const name of ['ShrId', 'EntryId', 'EntryType', 'FocalSubject', 'SubjectIsThirdPartyFlag', 'Narrative',
+        'Informant', 'Author', 'AssociatedEncounter', 'OriginalCreationDate', 'LastUpdateDate', 'Language']) {
+        entryFields['shr.base'][name] = true;
+      }
+    }
+    for (const de of this._expanded.dataElements.all) {
+      // TODO: What are the rules for inheriting values from multiple parents? Currently if a value or field matches any parent then it is considered inherited.
```
Comment (cmoesel) [inline lib/expand.js]: That seems like the right approach to me.

File: lib/expand.js [hunk 2]
```diff
@@ -43,6 +43,107 @@ class Expander {
...
+      if (de.basedOn && de.basedOn.length) {
+        const parents = [];
+        for (let i = 0; i < de.basedOn.length; i += 1) {
```
Comment (cmoesel) [inline lib/expand.js]: No big deal, but why `i += 1` rather than the more traditional `i++`?

Comment (cmoesel) [reply lib/expand.js]: Or better yet: `for (const basedOn of de.basedOn)`?

File: lib/expand.js [hunk 3]
```diff
@@ -43,6 +43,107 @@ class Expander {
...
+            logger.error(`Could not find based on element %s for child element %s.`, basedOn, de.identifier);
+          } else {
+            parents.push(parent);
+          }
+        }
+        if (de.value) {
+          let inheritance = null;
+          for (const parent of parents) {
+            if (parent.value) {
+              // TODO: This will flag TBD parent values as overridden. Is that the right choice? -jgibson
```
Comment (cmoesel) [inline lib/expand.js]: No, I think that we should treat TBD values as if they aren't there at all.  So, the child isn't overriding it until the someone makes the TBD real.  Unless you have a use case for why we would want it overridden when the parent is a TBD.

**Review by cmoesel — CHANGES_REQUESTED** [review]
Looks pretty good!  Only one serious suggestion (regarding TBD).  Thanks!

File: lib/expand.js (hunk 2 above)
Comment (jgibson) [reply lib/expand.js]: I usually use `i += 1` because one of my CS professors had a dislike for `i++`. However, it's moot because `for (const basedOn of de.basedOn)` is better anyways.

File: lib/expand.js (hunk 3 above)
Comment (jgibson) [reply lib/expand.js]: Done. TBD children with identical text values are still flagged as inherited.

**Comment by jgibson** [comment]
OK, I've made the changes requested and added some tests for TBD (they weren't being tested at all prior to this). I also found a bug where overridden TBD values weren't being expanded correctly. The only corner case that I can think of that is untested is an IncompleteValue overriding a TBD, but that seems fairly outlandish.

The expander also complains about things like missing cardinality on TBD values, which strikes me as overly pedantic if we want to consider TBD things as not really existing. However, given that we're looking to replace shr-models (and probably shr-expand, as well), it's probably not worth changing.

**Review by cmoesel — APPROVED** [review]
